src/audio/voice_internal.c

```diff
@@ -1,42 +1,367 @@
+    "PA", "HI", "BI", "PI", "HU", "BU", "PU", "HE", "BE", "PE", "HO", "BO", "PO", "MA", "MI", "  ",
+    "MU", "ME", "MO", "ya", "YA", "yu", "YU", "yo", "YO", "RA", "RI", "RU", "RE", "RO", "wa", "WA",
+    "WI", "WE", "WO", "NN", "VU", "ka", "ke", "",   "",   "",   "",   "",   "",   "",   "",   "",
+};
+
+const char D_801E0EC0[] = "Error %d\n";
+const char D_801E0ECC[] = "NAI_VRS:osVoiceClearDictionary %d\n";
+const char D_801E0EF0[] = "NAI_VRS:dict error! (%d-%d %s)\n";
+const char D_801E0F10[] = "NAI_VRS:Ina_SetVruGain Error!\n";
+const char D_801E0F30[] = "NAI_VRS:mask on  %d\n";
+const char D_801E0F48[] = "NAI_VRS:mask off %d\n";
+const char D_801E0F60[] = "NAI_VRS:answer No.:%d Dist:%d Warn:%04X Level:%5d SN:%5d (Num:%d)\n";
+const char D_801E0FA4[] = "NAI_VRS:error !! (ANS_MAX:%d DIST:%d WARNING:%04X LEVEL:%5d SN:%5d)\n";
+
+s32 func_801A51F0(s32 errorCode) {
+    switch (errorCode) {
+        case CONT_ERR_NO_CONTROLLER:
+        case CONT_ERR_CONTRFAIL:
+        case CONT_ERR_INVALID:
+        case CONT_ERR_DEVICE:
+        case CONT_ERR_VOICE_MEMORY:
+        case CONT_ERR_VOICE_WORD:
+        case CONT_ERR_VOICE_NO_RESPONSE:
             return -1;
+
         default:
             return 0;
     }
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/voice_internal/func_801A5228.s")
+s32 func_801A5228(OSVoiceDictionary* dict) {
+    OSMesgQueue* serialEventQueue;
+    s32 errorCode;
+    u8 numWords;
+    u8 i;
+
+    sVoiceContext.mode = VOICE_MODE_0;
+    sVoiceContext.data = NULL;
+    sVoiceContext.distance = 1000;
+    sVoiceContext.answerNum = 5;
+    sVoiceContext.warning = 0;
+    sVoiceContext.dict = dict;
+
+    numWords = dict->numWords;
+
+    serialEventQueue = PadMgr_VoiceAcquireSerialEventQueue();
+    errorCode = osVoiceClearDictionary(&gVoiceHandle, numWords);
+    PadMgr_VoiceReleaseSerialEventQueue(serialEventQueue);
+
+    if (errorCode != 0) {
+        return errorCode;
+    }
+
+    for (i = 0; i < (((numWords - 1) / 8) + 1); i++) {
+        sVoiceMaskPattern[i] = 0;
+    }
+
+    for (i = 0; i < numWords; i++) {
+        serialEventQueue = PadMgr_VoiceAcquireSerialEventQueue();
+        errorCode = osVoiceSetWord(&gVoiceHandle, (u8*)&dict->words[i]);
+        PadMgr_VoiceReleaseSerialEventQueue(serialEventQueue);
+
+        if (func_801A51F0(errorCode) != 0) {
+            func_801A5A1C((s8*)&dict->words[i]);
+        }
+    }
+
+    return errorCode;
+}
+
+OSVoiceData* func_801A5390(void) {
+    OSVoiceData* voiceData;
+    OSMesgQueue* serialEventQueue;
+
+    voiceData = sVoiceContext.data;
+    sVoiceContext.data = NULL;
+
+    serialEventQueue = PadMgr_VoiceAcquireSerialEventQueue();
+    osVoiceStartReadData(&gVoiceHandle);
+    PadMgr_VoiceReleaseSerialEventQueue(serialEventQueue);
+
+    return voiceData;
+}
+
+// Unused
+OSVoiceDictionary* AudioVoice_GetVoiceDict(void) {
+    return sVoiceContext.dict;
+}
+
+void func_801A53E8(u16 distance, u16 answerNum, u16 warning, u16 voiceLevel, u16 voiceRelLevel) {
```

Comment:
Maybe something like `SetVoiceContext`?

Reply:
It only sets part of `VoiceContext` though, just the 
```c
/* 0x0C */ u16 distance;
/* 0x0E */ u16 answerNum;
/* 0x10 */ u16 warning;
/* 0x12 */ u16 voiceLevel;
/* 0x14 */ u16 voiceRelLevel;
```
part, and it's still not clear how this data differs from what the `/* 0x08 */ OSVoiceData* data;` in voice data contains, which has the same data

Reply:
Ah, I missed that it was only part of it. Never mind then.